### PR TITLE
[WIP] Connecting JTAG for hardware debugging

### PR DESCRIPTION
**Pull request by hamishwillee**
Docs on how to connect JTAG for hardware debugging. I'm still hoping for at least one real screenshot showing a Pixhawk 4 connected by single cable to a jlink edu mini.

The aim is that you should be able to come with any pixhawk board and know what pins you need to connect on jlink edu mini to different pixhawk series boards. Info for other boards a "nice to have" but not necessarily vital in version 1.

@davids5 Many open questions. The aim is that

### CODE REVIEW

File: en/debug/gdb_debugging.md [hunk 1]
```diff
@@ -69,10 +69,25 @@ The numbers in the graph will all be zero, because they are in MB. Just look at
...
 
+## Hard Fault Debugging
 
-## Debugging Hard Faults in NuttX
+A hard fault is a state when the operating system detects that it has no valid instructions to execute.
+This is typically the case when key areas in RAM have been corrupted.
 
-A hard fault is a state when the operating system detects that it has no valid instructions to execute. This is typically the case when key areas in RAM have been corrupted. A typical scenario is when incorrect memory access smashed the stack and the processor sees that the address in memory is not a valid address for the microprocessors's RAM.
+
+### Video
```
Comment (hamishwillee) [inline en/debug/gdb_debugging.md]: FYI, this is just a "quick and dirty" way to get the video on hard fault debugging into docs. This section could certainly be revisited.

File: en/debug/jtag_pins.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+
+## Pixhawk Series
+
+[Pixhawk series](https://docs.px4.io/master/en/flight_controller/pixhawk_series.html) boards provide access to JTAG through the **Pixhawk Debug Port**.
+The physical connector and pin order for this port varies across Pixhawk FMU versions.
+
+> **Note** The **Pixhawk Debug Port** also exposes a UART for access to the [System Console](../debug/system_console.md) (a console that provides low-level access to the system, debug output and analysis of the system boot process).
+
+<!--
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: @davids5 
1. Do FMUv2 and 3 have debug ports?
2. If so, then what connector do they use (Hirose DF13), and what is the layout?

File: en/debug/jtag_pins.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+2 (blk) | CONSOLE TX (OUT) | +3.3V
+3 (blk) | CONSOLE RX (IN) | +3.3V
+4 (blk) | SWDIO | +3.3V
+5 (blk) | SWCLK | +3.3V
+6 (blk) | GND | GND
+
+
+### FMUv5X
+
+FMUv5X will have a [10-pin JST SUR](https://order.jst-mfg.com/InternetShop/app/index.php?back=2&product=53574000&jgcd=0#showProductDetail).
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: @davids5 Is this the type of FMUv5x debug port connector that is being proposed (see link: https://order.jst-mfg.com/InternetShop/app/index.php?back=2&product=53574000&jgcd=0#showProductDetail)

File: en/debug/jtag_pins.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+It adds JTAG pins: SWO, nRST. TRACECLK, TRACED0 Pins 7 and 8 are GPIOS to gate timing or 1 wire trace (full trace is on).
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: > ... TRACECLK, TRACED0 Pins 7 and 8 are GPIOS to gate timing or 1 wire trace (full trace is on).

What does this mean "full trace is on"? My guess is that SW is not "full trace" and that these new fields provide additional trace capabilities? 
Or to put it another way, will someone who knows anything about this topic know what you meant by that line?

File: en/debug/jtag_pins.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+
+## Hex Cube
+
+Cube debug port information can be found here: [Debug ports](http://docs.px4.io/master/en/flight_controller/pixhawk-2.html#debug-ports).
+
+If pulled apart the Cube has a 6 pin JST SUR connector (marked as J10) for both FMU and IO debugging.
+
+<!--
+Does the numbering match all the other 6 pin port examples.
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: Thanks for the info about cube FMU ports. 
- What label is used for the IO port?
- Is the in numbering the same as for other FMUv3 debug ports?

File: en/debug/jtag_pins.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+What label is the IO port marked as?
+So obviously it is much easier to just connect to the carrier board port. Who would know what interface that is. PhilipR?
+-->
+
+
+## Kakute F7
+
+There are Test points on the board.
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: @davids5. So you're saying that the board does support this. NOthing shows up on an internet serach. Any ideas about how I find out what pads do SWO etc?

File: en/debug/jtag_pins.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+UART3 RX and TX are configured for System console. 
+
+
+## Hovergames RDDRONE-FMUK66
+
+The Hovergames RDDRONE-FMUK66 has a similar debug interface to Pixhawk (it supports an additional JTAG pin - nRST) and uses a JST-GH connector). 
+The debug interface is documented here: [RDDRONE-FMUK66 Debug interface](https://nxp.gitbook.io/hovergames/rddrone-fmuk66/connectors/debug-interface-dcd-lz)
+
+![Debug RDDRONE-FMUK66 using Jlink and debug interface](../../assets/debug/jlink_hovergames.png)
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: @davids5 This is your diagram. What is the orange thing in the middle and where does that connection go? 

Is this just confusing? ie could they just wire up the lines directly?

File: en/debug/jtag_pins.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+<!-- what am I seeing here? Ie the jlink and the FC are obvious. 
+But what is the orange connector between them, and what is the connector out of the center of them doing? -->
+
+
+<!--
```
Comment (hamishwillee) [inline en/debug/jtag_pins.md]: @davids5 This is somewhat an aside. When do you use each debug port? ie is it that you expect people to know exactly what is loaded into FMU vs IO and process is same, for each, you just connect to a different port?

File: en/debug/jtag_pins.md (hunk 2 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: Yes. Use ARM 10 pin JTAG connector and a Special cable to get the console.  DF13 to  FTDI 3 wires. 
image

File: en/debug/jtag_pins.md (hunk 3 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: It is really not something ANYONE wants to assemble from parts.

Use a cable like this .  https://www.digikey.com/product-detail/en/jst-sales-america-inc/A10SUR10SUR32W152B/455-4015-ND/6708452

image

12:00  6 pin SUR
09:00 6 pin SH
6:00   10 pin SUR
image

File: en/debug/jtag_pins.md (hunk 4 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: Google ARM ETM

File: en/debug/jtag_pins.md (hunk 5 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: Ask HEX

File: en/debug/jtag_pins.md (hunk 6 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: Ask @bkueng for Photo I do not have one. @jinger26 can I get one?

File: en/debug/jtag_pins.md (hunk 7 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: Debugging means you have a problem. Making you own wiring adds a whole set of  potential faults in your fault tree.  == MORE PROBLEMS. Therefore debugging equipment needs to be rock solid and stable.  NXP, PX4, NscDG recognized this made adapter boards. 

image

NXP
image

Dronecode 
image

NscDG 
image

File: en/debug/jtag_pins.md (hunk 8 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: IO - load new bootloader or IO Firmware and debug IO (This is rare)
FMU load new bootloader or FMU Firmware and debug OR Just for the Console. (All the time for Developer)

File: en/debug/jtag_pins.md (hunk 7 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: The roll your own header uses cables like 
image
 for devices that have special connectors or even just PAD

**Comment by davids5** [comment]
>I'm still hoping for at least one real screenshot showing a Pixhawk 4 connected by single cable to a jlink edu mini.
image

The JTAG cable is long and tucked under where it is plugged into the FMU 

The lower one it connected to the IO. 

Tip:Leave the cables ALWAYS connected to the target!

File: en/debug/jtag_pins.md (hunk 2 above)
Comment (hamishwillee) [reply en/debug/jtag_pins.md]: @davids5 Thanks. Re the ARM 10pin, that gets mentioned in [docs here](https://docs.px4.io/master/en/flight_controller/pixhawk.html#parts--housings) but where is the actual connector on the board? It looks like you might have to rip off the cover and it is the empty space next to the GPS port?

**Comment by davids5** [comment]
Yes. The case has to be removed to access the connector for the FMU and IO.
And in most cases you have to solder on the connectors

File: en/debug/jtag_pins.md (hunk 2 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: @hamishwillee Yes. The case has to be removed to access the connector for the FMU and IO.
In most cases you have to solder on the connectors

FMU is in center (highlighted yellow)
IO is next to the PWM

image

File: en/debug/jtag_pins.md (hunk 6 above)
Comment (bkueng) [reply en/debug/jtag_pins.md]: I'm not aware that the KakuteF7 exposes JTAG.
The board with pics is here: https://shop.holybro.com/kakute-f7_p1104.html (I can make one if there's anything specific).

Comment (davids5) [reply en/debug/jtag_pins.md]: SWDIO -> TP3
SWCLK -> TP2
@bkueng can you see them and highlight in pic?

Comment (davids5) [reply en/debug/jtag_pins.md]: Pin 72 on CPU -> TP2->SWCLK
Pin 76 on CPU -> TP3->SWDIO
image

**Comment by dagar** [comment]
Continuing to refer to this as JTAG might be causing more confusion than it's worth. 

`JTAG -> SWD`

Then maybe a single mention near the top about what JTAG is and that we're actually using SWD.

**Comment by davids5** [comment]
a UART is a physical interface as is a JTAG interface, the mode you use it in relates to the protocol 

Serial Wire Debug (SWD) is a two-wire protocol for accessing the ARM debug interface. It is part of. the ARM Debug Interface Specification v5 and is an alternative to **JTAG. The physical layer of SWD.**

**Comment by dagar** [comment]
I'm not sure I follow. Are you suggesting calling the physical interface JTAG everywhere or just the specific JTAG connector (20 pin) to the debugger? 

From the ARM Debug Interface Specification v5 (ADIv5) - 
image

I'm suggesting we shouldn't be referring to the physical interface to the flight controller as JTAG (page title + most of the mentions).

> PX4 runs on hardware that exposes a JTAG interface in SW mode

or this

> PX4 requires the following pins for JTAG.



This is what we have for the pixhawk debug connector documentation.

image

**Comment by davids5** [comment]
@dagar 

Call it what it is on each board of the FMUvx series.  Then AKA it. To goal is for all leaves of experience to find this information.  Not calling it a JTAG interface obfuscates the capabilities an Embedded System Engineer needs to know exits.   Calling it Pixhawk Debug is fine but say what it is know as in the industry. This will educate a novice. 

Then Per board show the interface and where to find it for the IO and FMU.

FMUv2
image

FVMUv4 6 Pin 
  
image

FVMUv5 6 Pin 
image

FVMUv5 10 Pin 
image

**Comment by hamishwillee** [comment]
> Continuing to refer to this as JTAG might be causing more confusion than it's worth. JTAG -> SWD
  Then maybe a single mention near the top about what JTAG is and that we're actually using SWD.

I'm not sure what the conclusion exactly what above, but I will certainly make a clear statement.

I don't understand what the difference between FMUv4 and FMUv5 6 pin is? Does it matter to a user? Both are 6 pin with same ordering.  

Are we missing something for FMUv3 pro?

The 6 pin JST GH is referred to as the Pixhawk Debug Connector. What can I use to refer to the ten pin version?

**Comment by hamishwillee** [comment]
@davids5 Thanks for https://github.com/PX4/Devguide/pull/893#issuecomment-554358128 - Great pic, but no one can get this. It isn't for sale. 

So what I think we need is custom  wiring example showing Pixhawk 4 split out into console access and SWD - ideally connected to a Jlink edu mini. For a start this will show me if you "really" just connect the two SWD wires, or also need the Vref and GND too.

People like you need to be able to debug everything and anything. Most people just want a recommendation for a system that will do most things for them. For that a dronecode JST GH on one end and jlink edu mini on the other is a good choice right.

The only widely available option then is dronecode probe, because it works on the pixhawk debug port. So we could push that.

or to put it another way, what should we push for the everyman?

File: en/debug/jtag_pins.md (hunk 2 above)
Comment (hamishwillee) [reply en/debug/jtag_pins.md]: @davids5 How are these numbered? My assumption is that the little square indicates 1, and they are numbered 1-5, 6-10, with each pin being ordered as shown on the left.

image

File: en/debug/jtag_pins.md (hunk 6 above)
Comment (hamishwillee) [reply en/debug/jtag_pins.md]: I'm assuming @bkueng is still going to be marking this up.

File: en/debug/jtag_pins.md (hunk 2 above)
Comment (davids5) [reply en/debug/jtag_pins.md]: Googleing ARM 10 pin. https://www.google.com/search?client=ms-android-tmus-us-revc&sxsrf=ACYBGNRS_6Ne8zkJtNCMk_LBCL5yBHUWDw%3A1574212891193&ei=G5XUXdeZC5L--gSVxIigAw&q=arm+10+pin+pin+out&oq=arm+10+pin+pin+out&gs_l=mobile-gws-wiz-serp.3..0i13i30j0i8i13i30l2j33i160.4262.7876..8527...0.1..0.258.1329.0j6j2......0....1.........0i71j0j0i67j0i22i30j33i22i29i30j33i299.YW9MxSQ7bFE#imgdii=StG8YrFZIHFmmM:&imgrc=YqJDfEDkf8Uf_M:

**Comment by davids5** [comment]
> @davids5 Thanks for [#893 (comment)](https://github.com/PX4/Devguide/pull/893#issuecomment-554358128) - Great pic, but no one can get this. It isn't for sale.
> 
> So what I think we need is custom wiring example showing Pixhawk 4 split out into console access and SWD - ideally connected to a Jlink edu mini. For a start this will show me if you "really" just connect the two SWD wires, or also need the Vref and GND too.
> 
> People like you need to be able to debug everything and anything. Most people just want a recommendation for a system that will do most things for them. For that a dronecode JST GH on one end and jlink edu mini on the other is a good choice right.
> 
> The only widely available option then is dronecode probe, because it works on the pixhawk debug port. So we could push that.
> 
> or to put it another way, what should we push for the everyman?

All 4 wires are needed.

**Comment by hamishwillee** [comment]
Some updates, but this is still a mess. 

I have "some idea" of how to progress this.
1. For every board we document we should make sure that we provide proper information alongside that board so that people can work out what the debug interfaces are, and what cables and connectors they need. 
   - Started [doc to help get that info](https://docs.google.com/document/d/17o0AcHnj3tRytZ9Q7PH-juFvh_bmplI6j13MhRhZBdU/edit#)  as suggested by David.
   - Started adding relevant info in the boards we know: https://github.com/PX4/px4_user_guide/pull/616)
1. Then in "this" doc I want to talk about representative examples. Ie end to end cable example for Pixhawk 4 (or anything that uses the Pixhawk debug port. This should be enough for people to then work out how to do other boards.

File: en/debug/jtag_pins.md (hunk 6 above)
Comment (bkueng) [reply en/debug/jtag_pins.md]: Ah, magic. Thanks David.
kakutef7_swdio

Comment (davids5) [reply en/debug/jtag_pins.md]: @bkueng Nice! We should also show GND   and the VDD_3V3 
image

**Comment by davids5** [comment]
>I don't understand what the difference between FMUv4 and FMUv5 6 pin is?

None for late model production FMUv4 it is a JST SH SM06B-SRSS-TB(LF)(SN) https://www.digikey.com/product-detail/en/jst-sales-america-inc/SM06B-SRSS-TB(LF)(SN)/455-1806-1-ND/926877

Some early FMUv4 had a JST GH before the connector was called out on the page (circled in red) This This happened because the spec said FMU now use JST GH connector instead of DF-13. 

> Does it matter to a user? Both are 6 pin with same ordering. 

Only if they have an old one. Do justice by the users and get pictures of the JST GH and SH series connectors and put them in a sections called "How to identity your connector Types" Add links to digkey. Do not get confused seeing Green or Black JST GH connectors they are not pico blades.

See https://pixhawk.org/pixhawk-connector-standard/

image

The 6 pin was spec-ed (it is a SH connector)   JST SH SM06B-SRSS-TB(LF)(SN) https://www.digikey.com/product-detail/en/jst-sales-america-inc/SM06B-SRSS-TB(LF)(SN)/455-1806-1-ND/926877

The 3 Pads were to get access 

The 10 Pin adds SWO and specs the 3 pads as pins 
JST SH  -SM10B-SRSS-TB(LF)(SN) https://www.digikey.com/product-detail/en/jst-sales-america-inc/SM10B-SRSS-TB-LF-SN/455-1810-1-ND/926881?utm_adgroup=Connectors%20&%20Interconnects

**Comment by hamishwillee** [comment]
Replaced by https://github.com/PX4/Devguide/pull/984

Thanks for all your help on this!

**Pull request closed without merging**